Add gift card validation endpoint

hlbraddock opened the pull request:
Add gift card validation endpoint.

https://jira.cru.org/browse/EVENT-913

⏺ Validate Endpoint

  Endpoint

  GET /api/v1/gift_cards/:certificate/validate

  Authentication

  Same as other API endpoints - requires an API token passed via:
  - Query parameter: ?access_token=<token>
  - HTTP header: Authorization: token <token> or Authorization: Bearer <token>

  Example Request

  GET /api/v1/gift_cards/ABC123/validate?access_token=your_api_token_here

  or

  GET /api/v1/gift_cards/ABC123/validate
  Authorization: token your_api_token_here

  ---
  Response Examples

  ✅ Success - Gift Card is Valid (200 OK)

  When the gift card passes all validations:
  ```
{
    "valid": true,
    "gift_card": {
      "id": 1,
      "issuance_id": 1,
      "batch_id": 1,
      "price": "25.00",
      "gift_card_type": "paid_half_price",
      "certificate": "ABC123",
      "expiration_date": "2025-12-31",
      "registrations_available": 10,
      "associated_product": "product_name",
      "gl_code": "12345-dept-project",
      "isbn": "978-1234567890",
      "created_at": "2024-11-26T09:21:31.000Z",
      "updated_at": "2024-11-26T09:21:31.000Z"
    }
  }
```

  ❌ Error - No Registrations Available (403 Forbidden)

  When registrations_available is 1 or less:
```
  {
    "valid": false,
    "error": "No registrations available",
    "status": 403
  }
```
  ❌ Error - Gift Card Expired (403 Forbidden)

  When expiration_date is in the past:
```
  {
    "valid": false,
    "error": "Gift card has expired",
    "status": 403
  }
```

  ❌ Error - Gift Card Not Found (404 Not Found)

  When the certificate doesn't exist:
```
  {
    "error": "No gift card found by that certificate",
    "status": 404
  }
```

  ❌ Error - Invalid/Missing Token (401 Unauthorized)

  When the API token is missing or invalid:
```
  {
    "error": "You either didn't pass in an access token, or the token you did pass in was wrong.",
    "status": "unauthorized"
  }
```
  ---
  Validation Rules

  The endpoint checks the following conditions in order:

  1. Gift card exists → 404 if not found
  2. API token is valid → 401 if invalid/missing
  3. Has registrations available → Must have > 1 registrations (403 if not)
  4. Not expired → expiration_date must be today or future (403 if expired in past)
    - Note: If expiration_date is null, it's considered valid (no expiration)

  If all validations pass → Returns 200 OK with valid: true

  ---
  Key Differences from PUT /api/v1/gift_cards/:certificate

  | Feature           | GET validate             | PUT update                      |
  |-------------------|--------------------------|---------------------------------|
  | Purpose           | Check if card is valid   | Use a registration              |
  | Modifies data     | ❌ No                     | ✅ Yes (decrements by 2)         |
  | Checks expiration | ✅ Yes                    | ❌ No                            |
  | Response format   | {valid: true/false, ...} | Just gift card JSON             |
  | Use case          | Pre-validate before use  | Actually consume a registration |

Commit:
Add gift card validation endpoint

andrewroth commented on spec/controllers/api/v1/gift_cards_controller_spec.rb:
```diff
@@ -38,6 +38,44 @@
+    end
+
+    it "returns valid: false when no registrations available" do
+      gift_card.update(registrations_available: 1)
```
Similarly here,
`let(:gift_card_1_registration) { create(:gift_card, certificate: "cert", registrations_available: 10) }`

Commit:
Add expiration to the gift card validation

Commit:
Changes Made

  Before (lines 41-110)

  - Used inline gift_card.update(...) calls within each test
  - Modified the same gift_card object multiple times
  - Less clear what each test is testing

  After (lines 41-110)

  - Added 6 let statements at context level (lines 42-47):
    - gift_card_with_registrations - Has 10 registrations available
    - gift_card_no_registrations - Has only 1 registration (invalid)
    - gift_card_expired - Expired yesterday
    - gift_card_expires_today - Expires today (valid)
    - gift_card_expires_future - Expires tomorrow
    - gift_card_no_expiration - No expiration date set
  - Each test uses the appropriate let variable instead of updating
  - Each gift card has a unique certificate to avoid conflicts
  - Tests are cleaner and more declarative - you can see what state is being tested just by looking at the variable name

  This follows the RSpec convention of using let statements to set up test data at the context level, making tests more maintainable and easier to understand.
